File: src/add_two_ints_client.cpp
#include <ros/ros.h>
#include <bingda_prictices/addtwoints.h>//包含服务消息头文件    

int main(int argc, char **argv)
{
  ros::init(argc, argv, "add_two_ints_client");//初始化ROS节点，节点名称为"add_two_ints_client"
  ros::NodeHandle n;//创建一个节点句柄

  ros::ServiceClient client = n.serviceClient<bingda_prictices::addtwoints>("add_two_ints");//创建一个服务客户端，服务类型为bingda_prictices::addtwoints，服务名称为"add_two_ints"
  bingda_prictices::addtwoints srv;//创建一个服务对象

  srv.request.a = atoll(argv[1]);//将命令行参数转换为整数，并赋值给请求消息的a字段
  srv.request.b = atoll(argv[2]);//将命令行参数转换为整数，并赋值给请求消息的b字段

  if (client.call(srv))//调用服务，如果调用成功
  {
    ROS_INFO("Sum: %ld", (long int)srv.response.sum);//输出响应消息中的sum字段
  }
  else//如果调用失败
  {
    ROS_ERROR("Failed to call service add_two_ints");//输出错误信息
    return 1;
  }

  return 0;
}

File: scripts/add_two_ints_client.py
#!/usr/bin/env python3
import sys
import rospy
from bingda_prictices.srv import addtwoints  # 导入我们自己定义

def add_two_ints_client(x, y):
    rospy.wait_for_service('add_two_ints')  # 等待服务可用
    try:
        add_two_ints = rospy.ServiceProxy('add_two_ints', addtwoints)  # 创建服务代理 #addtwoints是服务名称，addtwoints是服务类型
        resp = add_two_ints(x, y)  # 调用服务，传入参数x和y
        return resp.sum  # 返回结果
    except rospy.ServiceException as e:
        print("Service call failed: %s" % e)

if __name__ == "__main__":
    if len(sys.argv) != 3: # 检查命令行参数是否正确
        print("Usage: add_two_ints_client.py [x y]")
        sys.exit(1)
    x = int(sys.argv[1])  # 从命令行参数获取第一个整数
    y = int(sys.argv[2])  # 从命令行参数获取第二个整数
    print("Requesting %d + %d" % (x, y))
    result = add_two_ints_client(x, y)  # 调用客户端函数，获取结果
    print("%d + %d = %d" % (x, y, result))  # 打印结果

